src/garage/misc/process_samples_utils.py

```diff
@@ -0,0 +1,80 @@
+"""Utiliy functions for batch processing samples."""
+from dowel import tabular
+import numpy as np
+
+from garage import log_performance, TrajectoryBatch
+from garage.misc import tensor_utils
+
+
+def preprocess_samples_for_policy_gradient(itr, paths, max_path_length,
+                                           baseline, discount, env_spec,
+                                           episode_reward_mean):
+    """Return processed sample data based on the collected paths.
+
+    Args:
+        itr (int): Iteration number.
+        paths (list[dict]): A list of collected paths.
+        max_path_length (int): Maximum length of a single rollout.
+        baseline(garage.np.baselines.Baseline): Baseline for GAE
+            (Generalized Advantage Estimation).
+        discount (float): Environment reward discount.
+        env_spec (garage.envs.EnvSpec): Environment specification.
+        episode_reward_mean(collections.deque): Deque to record episode
+            reward mean metric
+
+    Returns:
+        dict: Processed sample data, with key
+            * average_return: (float)
+
+    """
+    baselines = []
+    returns = []
+
+    if hasattr(baseline, 'predict_n'):
+        all_path_baselines = baseline.predict_n(paths)
+    else:
+        all_path_baselines = [baseline.predict(path) for path in paths]
+
+    for idx, path in enumerate(paths):
+        # baselines
+        path['baselines'] = all_path_baselines[idx]
+        baselines.append(path['baselines'])
+
+        # returns
+        path['returns'] = tensor_utils.discount_cumsum(path['rewards'],
+                                                       discount)
+        returns.append(path['returns'])
+
+    obs = [path['observations'] for path in paths]
+    obs = tensor_utils.pad_tensor_n(obs, max_path_length)
+
+    actions = [path['actions'] for path in paths]
+    actions = tensor_utils.pad_tensor_n(actions, max_path_length)
+
+    rewards = [path['rewards'] for path in paths]
+    rewards = tensor_utils.pad_tensor_n(rewards, max_path_length)
+
+    agent_infos = [path['agent_infos'] for path in paths]
+    agent_infos = tensor_utils.stack_tensor_dict_list([
+        tensor_utils.pad_tensor_dict(p, max_path_length) for p in agent_infos
+    ])
+
+    env_infos = [path['env_infos'] for path in paths]
+    env_infos = tensor_utils.stack_tensor_dict_list(
+        [tensor_utils.pad_tensor_dict(p, max_path_length) for p in env_infos])
+
+    valids = [np.ones_like(path['returns']) for path in paths]
+    valids = tensor_utils.pad_tensor_n(valids, max_path_length)
+
+    undiscounted_returns = log_performance(
```

Comment:
can you move this into the algo as well?

Reply:
> can you move this into the algo as well?

Sure!

Reply:
> can you move this into the algo as well?

Sorry, I am actually a little confused here. Do you mean moving line 66-69 into the algo? Or move everything starting from line 69? It seems to me that everything up to line 67 is "preprocessing", and codes starting from line 69 are actually processing the samples. 

Thanks in advance for the clarification!

Reply:
I am suggesting that the call to `log_performance` is really critical to all algorithms functioning correctly (it is actually calculating the performance of the algorithm) and should be moved into the algos.